vere/pier.c

```diff
@@ -846,160 +924,220 @@ _pier_disk_load_commit(u3_pier* pir_u,
 }
 
 /* _pier_boot_vent(): create and enqueue boot sequence
+**
+**  per cgy:
+**    this new boot sequence is almost, but not quite,
+**    the right thing.  see new arvo.
 */
 static void
 _pier_boot_vent(u3_pier* pir_u)
 {
-  c3_w inx_w = 1;
+  //  bot: boot formulas
+  //  mod: module ova
+  //  use: userpace ova
+  //
+  u3_noun bot, mod, use;
 
-  if ( !u3A->sys ) {
-    fprintf(stderr, "boot: loading pill %s\r\n", pir_u->sys_c);
+  //  extract boot formulas and module/userspace ova from pill
+  //
+  {
+    u3_noun pil_p, pil_q, pil_r;
+    u3_noun pro;
 
-    u3A->sys = u3m_file(pir_u->sys_c);
-  }
+    c3_assert( c3y == u3du(pir_u->pil) );
 
-  {
-    u3_noun lal = u3ke_cue(u3k(u3A->sys));
+    if ( c3y == u3h(pir_u->pil) ) {
+      u3x_trel(pir_u->pil, 0, &pil_p, &pil_q);
+    }
+    else {
+      u3x_qual(pir_u->pil, 0, &pil_p, &pil_q, &pil_r);
+    }
 
-    /* this new boot sequence is almost, but not quite,
-    ** the right thing.  see new arvo.
-    */
-    {
-      u3_noun who = u3i_chubs(2, pir_u->who_d);
-      u3_noun bot, mod, fil;
+    pro = u3m_soft(0, u3ke_cue, u3k(pil_p));
 
-      u3r_trel(lal, &bot, &mod, &fil);
-      pir_u->but_d = 0;
+    if ( 0 != u3h(pro) ) {
+      fprintf(stderr, "boot: failed: unable to parse pill\r\n");
+      exit(1);
+    }
 
-      /* insert boot sequence directly
-      */
-      {
-        u3_noun seq = u3k(bot);
-        {
-          u3_noun all = seq;
-
-          pir_u->but_d += u3kb_lent(u3k(all));
-          while ( all ) {
-            _pier_insert(pir_u, 0, u3k(u3h(all)));
-            inx_w++;
-            all = u3t(all);
+    u3x_trel(u3t(pro), &bot, &mod, &use);
+    u3k(bot); u3k(mod); u3k(use);
+
+    //  optionally replace filesystem in userspace
+    //
+    if ( c3y == u3h(pir_u->pil) ) {
+      if ( u3_nul != pil_q ) {
+        c3_w len_w = 0;
+        u3_noun ova = use;
+        u3_noun new = u3_nul;
+        u3_noun ovo;
+
+        while ( u3_nul != ova ) {
+          ovo = u3h(ova);
+
+          if ( c3__into == u3h(u3t(ovo)) ) {
+            c3_assert( 0 == len_w );
```

Comment:
We require that the pill contain at most one %into event?

Reply:
Good point. I had written this a little differently at first, always adding `u3t(pil_q)` as the first item in `new`, but this felt cleaner. I guess I should `c3_assert( 1 == len_w )` after the while loop. This is almost all just future-proofing, as both the `+brass` and `+solid` pill have just the `%into` in there userspace ova list.